pkg/worker/lifecycle.go

```diff
@@ -136,19 +136,30 @@ func (s *Worker) clearContainer(containerId string, request *types.ContainerRequ
 			}
 		}
 
-		s.containerInstances.Delete(containerId)
-		err = s.containerRepo.DeleteContainerState(containerId)
-		if err != nil {
-			log.Printf("<%s> - failed to remove container state: %v\n", containerId, err)
-		}
+		s.deleteContainer(containerId, err)
 
 		log.Printf("<%s> - finalized container shutdown.\n", containerId)
 	}()
 }
 
+func (s *Worker) deleteContainer(containerId string, err error) {
+	s.containerInstances.Delete(containerId)
+	err = s.containerRepo.DeleteContainerState(containerId)
+	if err != nil {
+		log.Printf("<%s> - failed to remove container state: %v\n", containerId, err)
+	}
+}
+
 // Spawn a single container and stream output to stdout/stderr
-func (s *Worker) RunContainer(request *types.ContainerRequest) error {
+func (s *Worker) RunContainer(request *types.ContainerRequest) (err error) {
 	containerId := request.ContainerId
+
+	defer func() {
```

Comment:
I don't think you need this, because there is already a call to clearcontainer upstream of this:

```
err := s.RunContainer(request)
				if err != nil {
					log.Printf("Unable to run container <%s>: %v\n", containerId, err)

					// Set a non-zero exit code for the container (both in memory, and in repo)
					exitCode := 1

					serr, ok := err.(*types.ExitCodeError)
					if ok {
						exitCode = serr.ExitCode
					}

					err := s.containerRepo.SetContainerExitCode(containerId, exitCode)
					if err != nil {
						log.Printf("<%s> - failed to set exit code: %v\n", containerId, err)
					}

					s.containerLock.Unlock()
					s.clearContainer(containerId, request, exitCode)
					continue
				}
```

Reply:
ahh I see. removed